tls/test_tls_cert.py

```diff
@@ -401,3 +430,678 @@ def test_vhost_cert_selection(self):
+        # save the current config text
+        original_config = self.get_tempesta().config.defconfig
+
+        # Both 'wildcard' and 'private' certificates are provided
+        for sni, expected_cert in (
+                ('example.com', 'wildcard'),
+                ('public.example.com', 'wildcard'),
+                ('private.example.com', 'private'),
+        ):
+            with self.subTest(msg="Trying TLS handshake", sni=sni):
+                hs = TlsHandshake(verbose=self.verbose)
+                hs.sni = [sni]
+                hs._do_12_hs()
+                self.assertTrue(x509_check_cn(hs.cert, expected_cert))
+
+        self.reload_with_config(self.config_no_private_section)
+        # After Tempesta reload, 'wildcard' certificate are provided for all subdomains
+        for sni, expected_cert in (
+                ('example.com', 'wildcard'),
+                ('public.example.com', 'wildcard'),
+                ('private.example.com', 'wildcard'),
+        ):
+            with self.subTest(msg="Trying TLS handshake after config reload", sni=sni):
+                hs = TlsHandshake(verbose=self.verbose)
+                hs.sni = [sni]
+                hs._do_12_hs()
+                self.assertTrue(x509_check_cn(hs.cert, expected_cert))
+
+        self.reload_with_config(self.config_only_private_section)
+        # After Tempesta reload,
+        # 'wildcard' certificate is provided for 'private' section,
+        hs = TlsHandshake(verbose=self.verbose)
+        hs.sni = ['private.example.com']
+        hs._do_12_hs()
+        self.assertTrue(x509_check_cn(hs.cert, 'private'))
+
+        # and no certificate provided for removed 'wildcard' section subdomains
+        for sni in 'example.com', 'public.example.com':
+            with self.subTest(msg="Check 'unknown server name' warning after reload", sni=sni):
+                with self.assertRaisesRegex(tls.TLSProtocolError, 'UNRECOGNIZED_NAME'):
+                    hs = TlsHandshake(verbose=self.verbose)
+                    hs.sni = [sni]
+                    hs._do_12_hs()
+
+        # After Tempesta reload, certificates are provided as at the beginning of the test
+        self.reload_with_config(original_config)
+        for sni, expected_cert in (
+                ('example.com', 'wildcard'),
+                ('public.example.com', 'wildcard'),
+                ('private.example.com', 'private'),
+        ):
+            with self.subTest(
+                    msg="Trying TLS handshake after second config reload",
+                    sni=sni
+            ):
+                hs = TlsHandshake(verbose=self.verbose)
+                hs.sni = [sni]
+                hs._do_12_hs()
+                self.assertTrue(x509_check_cn(hs.cert, expected_cert))
+
+
+class TlsSNIwithHttpTable(tester.TempestaTest):
+    """
+    Test that show that non-TLS vhost could be accessed with certificate from
+    another section.
+    Test should fail after the fix.
+    """
```

Comment:
Could you please also add a test for `http_host_required` that it requires (should require) equal SNI and `Host`/`:authority`? This way we can just jump to https://github.com/tempesta-tech/tempesta/issues/1688 w/o additional tests development.

Reply:
@krizhanovsky 
I added `TlsSniWithHttpTableMultiFrang` and `TlsSniWithHttpTableMultiH2Frang` tests,
which demonstrate that `http_host_required` does not stop access to any HTTP section.
https://github.com/tempesta-tech/tempesta-test/blob/fab03912b2f058663da929d32758661fc1e83ffc/tls/test_tls_cert.py#L1118

Such tests will be convenient for development of https://github.com/tempesta-tech/tempesta/issues/1688 , or should they be rewritten?

Reply:
I think it's better to go with the usual way: write the tests as they should be and mask them by #1688 for now